File: pipeline/validate_data.py
"""
Step 3: Validate the raw JSON returned by OpenAI against ContractData schema.

Validation is fully deterministic — no AI involved.
Raises a clear error if the structure is wrong; prints a warning for missing fields.
"""

from pydantic import ValidationError
from schema.contract_schema import ContractData


def validate_contract_data(raw_json: dict) -> ContractData:
    try:
        data = ContractData(**raw_json)
    except ValidationError as e:
        raise ValueError(f"Extracted data failed schema validation:\n{e}") from e

    if data.missing_fields:
        print(f"      ! Fields not found by AI: {', '.join(data.missing_fields)}")

    return data

File: src/pipeline/validate_data.ts
import { ContractDataSchema, type ContractData } from '../schema/contract_schema'

export function validateContractData(rawJson: Record<string, unknown>): ContractData {
  const result = ContractDataSchema.safeParse(rawJson)

  if (!result.success) {
    throw new Error(`Extracted data failed schema validation:\n${result.error.message}`)
  }

  const data = result.data

  if (data.missing_fields.length > 0) {
    console.log(`      ! Fields not found by AI: ${data.missing_fields.join(', ')}`)
  }

  return data
}
